INDI/INDI/indiserver.c

```diff
@@ -838,46 +824,57 @@ static void shutdownDvr(DvrInfo *dp, int restart)
         /* close local named FIFOs */
         close(dp->wfd);
         close(dp->rfd);
-#       if 0
-        close(dp->efd);
-#       endif/*0*/
     }
 
 #ifdef OSX_EMBEDED_MODE
     fprintf(stderr, "STOPPED \"%s\"\n", dp->name);
     fflush(stderr);
 #endif
 
-    /* free memory */
-    free(dp->sprops);
-    free(dp->dev);
-    delLilXML(dp->lp);
+    /* free memory; ensure no double-free if stop interrupts restart */
+    if (dp->sprops) { free(dp->sprops); dp->sprops = NULL; }
+    if (dp->dev) { free(dp->dev); dp->dev = NULL; }
+    if (dp->lp) { delLilXML(dp->lp); dp->lp = NULL; }
 
     /* ok now to recycle */
     dp->active = 0;
-    dp->ndev   = 0;
 
     /* decrement and possibly free any unsent messages for this client */
-    while ((mp = (Msg *)popFQ(dp->msgq)) != NULL)
-        if (--mp->count == 0)
-            freeMsg(mp);
-    delFQ(dp->msgq);
+    if (dp->msgq)
+    {
+        while ((mp = (Msg *)popFQ(dp->msgq)) != NULL)
+        {
+            if (--mp->count == 0) { freeMsg(mp); }
+        }
+        delFQ(dp->msgq);
+        dp->msgq = NULL;
+    }
 
     if (restart)
     {
-        if (dp->restarts >= maxrestarts)
+        if (dp->restarts >= maxrestarts && maxrestarts > 0)
         {
-            fprintf(stderr, "%s: Driver %s: Terminated after #%d restarts.\n", indi_tstamp(NULL), dp->name,
-                    dp->restarts);
+            fprintf(stderr, "%s: Driver %s: Terminated after #%d restarts.\n"
+                   , indi_tstamp(NULL), dp->name, dp->restarts);
             // If we're not in FIFO mode and we do not have any more drivers, shutdown the server
             terminateddrv++;
-            if ((ndvrinfo - terminateddrv) <= 0 && !fifo.name)
-                Bye();
+            if ((ndvrinfo-terminateddrv) <= 0 && !fifo.name) { Bye(); }
         }
         else
         {
-            fprintf(stderr, "%s: Driver %s: restart #%d\n", indi_tstamp(NULL), dp->name, ++dp->restarts);
-            startDvr(dp);
+            /* Add to end of linked list of drivers to restart */
+            ppDvr ppdvr = &pRestarts;
+            while (*ppdvr) { ppdvr = &(*ppdvr)->pNextToRestart; }
+            *ppdvr = dp;
+            dp->pNextToRestart = NULL;     /* terminate linked list */
+
+            /* Prevent recylcing, set 10s delay until restart */
+            dp->active = 1;
+            dp->restartDelayus = 10 * Mus;
+
+            fprintf(stderr, "%s: Driver %s: scheduled for restart #%d in %lfs\n"
+                   , indi_tstamp(NULL), dp->name, ++dp->restarts
+                   , dp->restartDelayus / ((double)Mus));
```

Comment:
Restarts attempts occur at 0.1Hz i.e. occur after a 10s delay since the last communication failure.

This changed code sets up the DvrInfo data to implement that delay.